Community/Expire/automation/jsr223/python/community/expire/expire.py

```diff
@@ -103,127 +106,239 @@ def get_config(item_name):
+            timers[item_name].cancel()
+        return
 
     # Cancel the timer when the Item enters the cfg state.
     # Use unicode because there is no degree symbol in ASCII so we can handle
     # Number:Temperature Items.
-    if unicode(items[event.itemName]) == cfg["state"]:
-        if (event.itemName in timers
-                and not timers[event.itemName].hasTerminated()):
-            timers[event.itemName].cancel()
-            del timers[event.itemName]
+    if unicode(items[item_name]) == cfg["state"]:
+        if (timers[item_name] is not None
+                and not timers[item_name].hasTerminated()):
+            timers[item_name].cancel()
+        return
 
     # Create an expire timer when the Item differs from the end state.
+    expire.log.debug("Setting timer for '{}' with delay {}"
+                     .format(item_name, cfg["time"]))
+    t = (DateTime.now().plusDays(cfg["time"].days)
+                        .plusSeconds(cfg["time"].seconds)
+                        .plusMillis(int(cfg["time"].microseconds/1000)))
+    if (timers[item_name] is not None
+            and not timers[item_name].hasTerminated()):
+        timers[item_name].reschedule(t)
+    else:
+        timers[item_name] = ScriptExecution.createTimer(
+            t,
+            lambda: expired(ir.getItem(item_name),
+                            cfg["type"],
+                            cfg["state"],
+                            expire.log)
+        )
+
+
+def expire_load(event):
+
+    log = logging.getLogger("{}.Expire Load".format(LOG_PREFIX))
+    log.debug("Expire loading...")
+
+    # Keep track of items configured this pass
+    new_items = []
+
+    # Scan for items with valid expire config
+    for item_name in items:
+        cfg = get_config(item_name, log)
+        if cfg:
+            new_items.append(item_name)
+            timers[item_name] = timers.get(item_name, None) # don't clobber
+            log.debug("Expire configured for '{}' with timeout {} to {} '{}'"
+                      .format(item_name, cfg["time"],
+                      cfg["type"], cfg["state"]))
+
+    # Remove existing rule
+    if hasattr(expire, "UID"):
+        rules.remove(expire.UID)
+        delattr(expire, "triggers")
+        delattr(expire, "UID")
+
+    # Generate triggers
+    for item_name in new_items:
+        when("Item {} received update".format(item_name))(expire)
+
+    # Create expire rule
+    if hasattr(expire, "triggers"):
+        rule(
+            "Expire",
+            description=("Simulates the Expire1 binding, updating or commanding an "
+                         "Item after a configured amount of time."),
+            tags=["expire"]
+        )(expire)
+        if hasattr(expire, "UID"):
+            log.info("Expire loaded successfully")
+        else:
+            log.error("Failed to create Expire rule")
+    else:
+        log.info("Expire found no configured items")
+
+    # Drop items that no longer exist or have expire config
```

Comment:
"exist or no longer have an expire config" Without the "no longer" repeated the sentence is a little ambiguous.

Reply:
Agreed